pkg/entitlements/service.go

```diff
@@ -0,0 +1,112 @@
+package entitlements
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/canonical/identity-platform-admin-ui/internal/logging"
+	"github.com/canonical/identity-platform-admin-ui/internal/monitoring"
+	v1 "github.com/canonical/rebac-admin-ui-handlers/v1"
+	"github.com/canonical/rebac-admin-ui-handlers/v1/resources"
+	openfga "github.com/openfga/go-sdk"
+	"go.opentelemetry.io/otel/trace"
+)
+
+type V1Service struct {
+	ofga OpenFGAClientInterface
+
+	tracer  trace.Tracer
+	monitor monitoring.MonitorInterface
+	logger  logging.LoggerInterface
+}
+
+func (s *V1Service) ListEntitlements(ctx context.Context, params *resources.GetEntitlementsParams) ([]resources.EntitlementSchema, error) {
+	ctx, span := s.tracer.Start(ctx, "entitlements.V1Service.ListEntitlements")
+	defer span.End()
+
+	authModel, err := s.ofga.ReadModel(ctx)
+	if err != nil {
+		s.logger.Errorf("failed to read model: %v", err)
+		return nil, v1.NewUnknownError(fmt.Sprintf("failed to read the authorization model: %v", err))
+	}
+
+	entitlementSchemas := make([]resources.EntitlementSchema, 0)
+	for _, typeDef := range authModel.TypeDefinitions {
+		if typeDef.Metadata == nil {
+			continue
+		}
+
+		if params.Filter != nil && !strings.Contains(typeDef.Type, *params.Filter) {
+			continue
+		}
+
+		relations := *typeDef.GetMetadata().Relations
+		for relation, relationMetadata := range relations {
+			relationReferences := relationMetadata.GetDirectlyRelatedUserTypes()
+			receivers := buildReceivers(relationReferences)
+			entitlementSchema := resources.EntitlementSchema{
+				Entitlement:  relation,
+				EntityType:   typeDef.Type,
+				ReceiverType: receivers,
+			}
+
+			entitlementSchemas = append(entitlementSchemas, entitlementSchema)
+		}
+	}
```

Comment:
following up on point above, we should probably cache this so we don't repeat it every time it gets called